File: leetcode/DP/871.Minimum Number of Refueling Stops/871Minimum Number of Refueling Stops.cpp
class Solution {
public:
    int minRefuelStops(int target, int startFuel, vector<vector<int>>& stations) {
        int cur = startFuel;
        int stops = 0;
        int i = 0;
        priority_queue<int> q;
        while (true)
        {
            if (cur >= target) return stops;
            while (i<stations.size() && stations[i][0] <= cur)
                q.push(stations[i++][1]);
            if (q.empty()) break;
            cur += q.top(); q.pop();
            ++stops;
        }
        return -1;
    }
};

File: leetcode/DP/871.Minimum Number of Refueling Stops/871Minimum Number of Refueling Stops.py
class Solution:
    def minRefuelStops(self, target: int, startFuel: int, stations: List[List[int]]) -> int:
        # dp[j] 加油j次后最远能到哪
        # dp[0] = startFuel
        # 01背包问题
        dp = [0] * (len(stations)+1)
        dp[0] = startFuel
        for i in range(0, len(stations)):
            for j in range(i+1, 0, -1):
                if dp[j-1] >= stations[i][0]:
                    dp[j] = max(dp[j], dp[j-1]+stations[i][1])
       
        for i, dpp in enumerate(dp):
            if dpp >= target:
                return i
        return -1